sonar-go-to-slang/slang-checks/src/main/java/org/sonarsource/slang/checks/SliceEqualSimplificationCheck.java

```diff
@@ -0,0 +1,192 @@
+ * along with this program; if not, see https://sonarsource.com/license/ssal/
+ */
+package org.sonarsource.slang.checks;
+
+import java.util.List;
+import org.sonar.check.Rule;
+import org.sonarsource.slang.api.BinaryExpressionTree;
+import org.sonarsource.slang.api.BlockTree;
+import org.sonarsource.slang.api.FunctionDeclarationTree;
+import org.sonarsource.slang.api.FunctionInvocationTree;
+import org.sonarsource.slang.api.IdentifierTree;
+import org.sonarsource.slang.api.IfTree;
+import org.sonarsource.slang.api.IndexExpressionTree;
+import org.sonarsource.slang.api.LoopTree;
+import org.sonarsource.slang.api.ParameterTree;
+import org.sonarsource.slang.api.ReturnTree;
+import org.sonarsource.slang.api.Tree;
+import org.sonarsource.slang.api.Type;
+import org.sonarsource.slang.checks.api.CheckContext;
+import org.sonarsource.slang.checks.api.InitContext;
+import org.sonarsource.slang.checks.api.SlangCheck;
+
+@Rule(key = "S-slice-equal")
+public class SliceEqualSimplificationCheck implements SlangCheck {
+
+  private static final String MESSAGE = "Use slices.Equal() instead of custom slice comparison";
+
+  @Override
+  public void initialize(InitContext init) {
+    init.register(FunctionDeclarationTree.class, this::visitFunctionDeclaration);
+  }
+
+  private void visitFunctionDeclaration(CheckContext ctx, FunctionDeclarationTree function) {
+    if (isSliceEqualityFunction(function)) {
+      ctx.reportIssue(function.name(), MESSAGE);
+    }
+  }
+
+  private static boolean isSliceEqualityFunction(FunctionDeclarationTree function) {
+    List<ParameterTree> params = function.formalParameters();
+    if (params.size() != 2) {
+      return false;
+    }
+
+    ParameterTree param1 = params.get(0);
+    ParameterTree param2 = params.get(1);
+
+    // Semantic checks would be needed here to verify types.
+    // For example:
+    // if (!isSlice(param1.type()) || !isSlice(param2.type())) {
+    //   return false;
+    // }
+    // if (!areElementTypesComparableAndIdentical(param1.type(), param2.type())) {
+    //   return false;
+    // }
+    // if (!isBool(function.returnType())) {
+    //   return false;
+    // }
+
+    BlockTree body = function.body();
+    if (body == null || body.statementOrExpressions().size() != 3) {
+      return false;
+    }
+
+    List<Tree> statements = body.statementOrExpressions();
+    String p1Name = param1.identifier().name();
+    String p2Name = param2.identifier().name();
+
+    return isLengthCheck(statements.get(0), p1Name, p2Name) &&
+      isElementComparisonLoop(statements.get(1), p1Name, p2Name) &&
+      isReturnTrue(statements.get(2));
+  }
+
+  private static boolean isLengthCheck(Tree statement, String p1Name, String p2Name) {
+    if (!(statement instanceof IfTree)) {
+      return false;
+    }
+    IfTree ifTree = (IfTree) statement;
+    if (ifTree.elseClause() != null || !(ifTree.condition() instanceof BinaryExpressionTree)) {
```

Comment:
```suggestion
    if (ifTree.elseBranch() != null || !(ifTree.condition() instanceof BinaryExpressionTree)) {
```

After extracting from SLang some method names changed (the package also changed so it is not a breaking change in API)